### PR TITLE
Full site editor: load content in iframe

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Same as #21102, but for the full site editor.

See #20797.

> * There would be no admin CSS bleed at all. This is something we’ve been struggling with since the beginning.
> * There would be no need to simulate media queries, which is arguably technically more difficult than using an iframe.
> * Relative units like `(r)em` and `vw`/`vh` just work.
> * For a full site, a theme stylesheet can be just dropped in the editor without any adjustment. I think this is important as it makes the life of theme authors much easier.
> * It's possible to have one selection per window, so one in the admin and one in the content. This is useful for e.g. the link UI where the selection in the content can be kept while the selection is also in an input element (for the URL). Maybe be useful in other cases.

## How has this been tested?

The full site editor should work as before.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jasmussen** [comment]
This works _surprisingly_ well considering the technical achievement it is. It's virtually a non issue. 

I noticed two things:

Screenshot 2020-10-05 at 09 16 45

That shows a stock Chrome focus style around the iframe itself (.edit-site-visual-editor). It's inset 24px, which happens to match what I have in my editor style:

```
html {
    padding: 24px;
}
```

Here's a little inspector debugging:

inset

It's a little weird because the `iframe` element appears to be correctly positioned so it would seem that the iframe focus style should go edge to edge (where by the way we could give it a nicer focus style, once that is consistent with the rest of the interface).

Speaking of the focus style, it's supposed to inherit the spot color from the wp-admin theme, I'm using "Modern":

Screenshot 2020-10-05 at 09 24 32

CC: @youknowriad 

Finally perhaps more of a support question: I'm using a Google font, called "Mulish" in my theme and editor style. I've been enqueueing it like so:

```
add_action('admin_head', 'mulish_fonts');
```

That's almost certainly the wrong way to do it. However what would be the right way to do it?

Really astoundingly nice work here Ella!

**Comment by ellatrix** [comment]
@jasmussen Thanks for testing! Good question about the fonts. The font should be a dependency of the editor styles, and then it should (or should be able to) be added properly in the iframe too.

**Comment by ellatrix** [comment]
@jasmussen It looks like everything has to go through the `add_editor_style` API. The problem with that is that you can't add any dependencies? I'm not sure. Alternatively you could use `@import` but we should definitely make dependencies work.

**Comment by jasmussen** [comment]
I haven't had a chance to try quite yet. And I'd also assume it's something we can fix — I'm fine with having to do a little theme rejiggering to get my Google fonts to work.

**Comment by ellatrix** [comment]
@jasmussen The focus padding issues should be fixed now.

File: packages/edit-site/src/components/block-editor/index.js [hunk 1]
```diff
@@ -12,24 +18,136 @@ import {
...
+		doc.body.addEventListener( 'focus', onFocus );
+
+		return () => {
+			doc.body.removeEventListener( 'focus', onFocus );
+		};
+	}, [ onFocus ] );
+
+	useEffect( () => {
+		doc.body.className = bodyClassName;
+		doc.body.style.margin = '0';
```
Comment (nosolosw) [inline packages/edit-site/src/components/block-editor/index.js:51]: Why do we need this margin for? The reason I noticed is because [I'm using](https://github.com/nosolosw/global-styles-theme/blob/master/style.css#L16) the `margin: auto` trick to center content in the middle for test purposes and this body.margin messes up with it (note how tho content is not centered).

Screenshot from 2020-10-06 16-34-45

Comment (ellatrix) [reply packages/edit-site/src/components/block-editor/index.js:51]: I think we can remove it now. It was meant as a reset when I first started working on this, but themes should reset this.

Comment (ZebulanStanphill) [reply packages/edit-site/src/components/block-editor/index.js:51]: I'm really starting to think that using margins like this to handle default "alignment"/width in a theme was a bad idea. It hinders blocks from defining their own left/right margins on their root element (something I ran into while working on #21711), interferes with inner-block styles (I ran into this while working on #23222), and in general feels really fragile. (#25892, for example, currently has the Quote block up against the left side of the screen, similar to the situation in the screenshot @nosolosw posted.)

I wonder if we should consider making a recommendation against this styling method and avoid it in future default WP themes.

Comment (ellatrix) [reply packages/edit-site/src/components/block-editor/index.js]: I remove the inline style. Does it fix the issue, @nosolosw?

File: packages/block-editor/src/components/iframe/index.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+			if ( ! cssRules ) {
+				return;
+			}
+
+			const isMatch = Array.from( cssRules ).find(
+				( { selectorText } ) =>
+					selectorText &&
+					( selectorText.includes( `.${ className }` ) ||
+						selectorText.includes( `.wp-block` ) )
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/iframe/index.js:42]: I'm not sure how much of a problem this actually is, but couldn't `selectorText.includes( '.wp-block' )` end up detecting stuff like `.wp-blockz` or `.wp-block-abcde`?

Also, there's an unnecessary usage of backticks here instead of single quotes.

File: packages/block-editor/src/components/iframe/index.js [hunk 3]
```diff
@@ -0,0 +1,225 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, useEffect, createPortal, useRef } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { useResizeObserver } from '@wordpress/compose';
+
+const className = 'editor-styles-wrapper';
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/iframe/index.js:8]:
```suggestion
const CLASS_NAME = 'editor-styles-wrapper';
```

File: packages/block-editor/src/components/iframe/index.js [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+			if ( isMatch && ! doc.getElementById( ownerNode.id ) ) {
+				doc.head.appendChild( ownerNode.cloneNode( true ) );
+			}
+		} );
+	}, [] );
+}
+
+/**
+ * Bubbles some event types (keydown, keypress and dragover) to parent document
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/iframe/index.js:53]: Do we use Oxford commas?
```suggestion
 * Bubbles some event types (keydown, keypress, and dragover) to parent document
```

File: packages/components/src/popover/index.js [hunk 5]
```diff
@@ -88,13 +115,15 @@ function computeAnchorRect(
...
 		const bottomRect = bottom.getBoundingClientRect();
-		const rect = new window.DOMRect(
+		let rect = new window.DOMRect(
 			topRect.left,
 			topRect.top,
 			topRect.width,
 			bottomRect.bottom - topRect.top
 		);
 
+		rect = offsetIframe( rect, top.ownerDocument );
```
Comment (ZebulanStanphill) [inline packages/components/src/popover/index.js:125]: Why isn't this just:
```suggestion
		const rect = offsetIframe(
			new window.DOMRect(
				topRect.left,
				topRect.top,
				topRect.width,
				bottomRect.bottom - topRect.top
			),
			top.ownerDocument
		);
```

File: packages/components/src/popover/index.js [hunk 6]
```diff
@@ -42,6 +42,24 @@ const FocusManaged = withConstrainedTabbing(
  */
 const SLOT_NAME = 'Popover';
 
+function offsetIframe( rect, ownerDocument ) {
+	const { defaultView } = ownerDocument;
+	const { frameElement } = defaultView;
```
Comment (ZebulanStanphill) [inline packages/components/src/popover/index.js:47]:
```suggestion
	const { defaultView: { frameElement } } = ownerDocument;
```

File: packages/block-editor/src/components/iframe/index.js [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+			if ( cancelled ) {
+				event.preventDefault();
+			}
+		}
+
+		const eventTypes = [ 'keydown', 'keypress', 'dragover' ];
+
+		eventTypes.forEach( ( name ) => {
+			doc.addEventListener( name, bubbleEvent );
+		} );
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/iframe/index.js:96]: In general, there are a bunch of cases where `for...of` loops could be used instead of `[].forEach()`, though I'm not sure which is more performant... just that I think `for...of` is easier to read.
```suggestion
		for ( const name of eventTypes ) {
			doc.addEventListener( name, bubbleEvent );
		}
```

**Review by ZebulanStanphill — COMMENTED** [review]
Haven't actually tested the code yet, but here are some (mostly) stylistic suggestions.

File: packages/block-editor/src/components/iframe/index.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:42]: It could. It's actually meant to also detect `.wp-block-something` like `.wp-block-paragraph` etc. I'll update the comment.

File: packages/block-editor/src/components/iframe/index.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:53]: I don't care tbh :) I don't think we do consistently.

File: packages/block-editor/src/components/iframe/index.js (hunk 7 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:96]: Agreed.

File: packages/components/src/popover/index.js (hunk 6 above)
Comment (ellatrix) [reply packages/components/src/popover/index.js:47]: Both constants are needed.

Comment (ZebulanStanphill) [reply packages/components/src/popover/index.js:47]: Both are still usable in my suggestion, but I'm just nitpicking here, so it really doesn't matter. :stuck_out_tongue:

**Comment by jasmussen** [comment]
Took this for a spin today. It's REALLY close. And it's going to be so sweet to get this down. @MaggieCabrera and @scruffian you'll want to see this :) 

Most of the stuff I'd expect to break just works. It's just a really good experience. I did notice a few small things, though:

Screenshot 2020-12-10 at 13 01 23

There's a small resize handle on this field. I don't see it in the master branch, not sure what's up there.

As a potentially slightly bigger issue, we've got soem scrollbar shenanigans going on with some particularly complicated templates:

Screenshot 2020-12-10 at 13 04 11

As best I can tell what's going on here is this CSS:

```
.interface-interface-skeleton__content {
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    overflow: auto;
}
```

that accounts for the outermost scrollbar. 

Then there's the iframe itself, which scales to fit with these inline styles that are updated dynamically:

```
display: block;
width: 100%;
min-height: 100%;
height: 5142px;
```

After some debugging, I've narrowed it down to this:

```
.wp-block[data-align="full"], .wp-block.alignfull {
    max-width: 100vw!important;
    width: 100vw!important;
}
```

That appears to be output inline:

Screenshot 2020-12-10 at 13 20 11

I'm not sure what added it, whether it was my own theme, TwentyTwentyOne Blocks (which I also tested), or something else. But the tricky thing is that 100vw does not include scrollbars in their math, which means anything that's this wide ironically causes scrollbars to appear. 

It looks like the iframe isn't actually the container that scrolls the content, but that a parent container that holds the iframe actually scrolls the content and the iframe is then resized with the browser or content. I'm going to assume there's a really good reason for that, but it also seems to be causing issues when you don't have auto-hiding scrollbars.

File: lib/client-assets.php [hunk 8]
```diff
@@ -662,3 +662,40 @@ function gutenberg_extend_block_editor_settings_with_fse_theme_flag( $settings )
...
+
+	ob_start();
+
+	wp_styles()->done = array();
+	wp_styles()->do_items( $handles );
+	wp_styles()->done = $done;
+
+	$editor_styles = wp_json_encode( array( 'html' => ob_get_clean() ) );
+
+	echo "<script>window.__editorStyles = $editor_styles</script>";
```
Comment (youknowriad) [inline lib/client-assets.php:699]: Is there a better to do that without the new variable. Maybe by just reusing the "styles" property in the editor settings?

File: packages/block-editor/src/components/block-list/block-popover.js [hunk 9]
```diff
@@ -167,7 +167,9 @@ function BlockPopover( {
 		: 'top right left';
 	const stickyBoundaryElement = showEmptyBlockSideInserter
 		? undefined
-		: getScrollContainer( node ) || ownerDocument.body;
+		: getScrollContainer( ownerDocument.defaultView.frameElement ) ||
+		  getScrollContainer( node ) ||
+		  ownerDocument.body;
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/block-popover.js:172]: Do you think this could use a comment to explain the different levels of fallback?

File: packages/block-editor/src/components/block-list/block-selection-button.js [hunk 10]
```diff
@@ -223,6 +223,12 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
 
 				if ( navigateDown ) {
 					nextTabbable = focus.tabbable.findNext( blockElement );
+
+					if ( ! nextTabbable ) {
+						nextTabbable =
+							blockElement.ownerDocument.defaultView.frameElement;
+						nextTabbable = focus.tabbable.findNext( nextTabbable );
+					}
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/block-selection-button.js:231]: Is this the writing flow behavior? Did it move to the selection button? and what does this fallback do?

File: packages/block-editor/src/components/iframe/index.js [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+					( selectorText.includes( `.${ BODY_CLASS_NAME }` ) ||
+						selectorText.includes( `.${ BLOCK_PREFIX }` ) )
+			);
+
+			if ( isMatch && ! doc.getElementById( ownerNode.id ) ) {
+				doc.head.appendChild( ownerNode.cloneNode( true ) );
+			}
+		} );
+	}, [] );
+}
```
Comment (youknowriad) [inline packages/block-editor/src/components/iframe/index.js:51]: Is this meant for stylesheets from themes and plugins not using the "official" editor styles way of enqueuing things? Should we add a deprecated call when we find one of these?

File: packages/block-editor/src/components/iframe/index.js [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+ * Bubbles some event types (keydown, keypress, and dragover) to parent document
+ * document to ensure that the keyboard shortcuts and drag and drop work.
+ *
+ * Ideally, we should remove event bubbling in the future. Keyboard shortcuts
+ * should be context dependent, e.g. actions on blocks like Cmd+A should not
+ * work globally outside the block editor.
+ *
+ * @param {Document} doc Document to attach listeners to.
+ */
+function useBubbleEvents( doc ) {
```
Comment (youknowriad) [inline packages/block-editor/src/components/iframe/index.js:63]: Do you know in how many places this hook is needed? How confident are you that we'll be able to remove this quickly by doing the right refactors?

File: packages/block-editor/src/components/iframe/index.js [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+ * Positions the body element so that the resize listener works correctly. We're
+ * using an absolute position here because the resize listener doesn't seem to
+ * report shrinking when the position is relative, causing the iframe not to
+ * shrink when content is removed.
+ *
+ * @see https://github.com/FezVrasta/react-resize-aware#usage
+ *
+ * @param {Document} doc Document to set styles to.
+ */
+function useResizeListenerCompat( doc ) {
```
Comment (youknowriad) [inline packages/block-editor/src/components/iframe/index.js:140]: This one I have trouble understanding :P

File: packages/block-editor/src/components/iframe/index.js [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+ * @param {string}   head HTML to set as the head.
+ */
+function useHead( doc, head ) {
+	useEffect( () => {
+		doc.head.innerHTML =
+			// Body margin must be overridable by themes.
+			'<style>body{margin:0}</style>' +
+			'<style>.wp-block[data-align="full"],.wp-block.alignfull{max-width:100vw!important;width:100vw!important;}</style>' +
+			head;
+	}, [] );
```
Comment (youknowriad) [inline packages/block-editor/src/components/iframe/index.js:162]: Why are these styles necessary? Does it mean these should be included in the  "default" editor styles or something?

**Comment by MaggieCabrera** [comment]
Thanks for your work @jasmussen I'm very excited about this!
> 
> There's a small resize handle on this field. I don't see it in the master branch, not sure what's up there.
> 

Looks like it's not on master and we are overriding this:

```
.block-editor-default-block-appender textarea.block-editor-default-block-appender__content {
  resize: none;
}
```

with inline styles:

Screenshot 2020-12-10 at 13 36 06

File: packages/block-editor/src/components/iframe/index.js [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+	useEffect( () => {
+		setDocumentIfReady( ref.current.contentDocument );
+	}, [] );
+
+	function setRef( newRef ) {
+		ref.current = newRef;
+
+		if ( newRef ) {
+			setDocumentIfReady( newRef.contentDocument );
```
Comment (youknowriad) [inline packages/block-editor/src/components/iframe/index.js:195]: It seems this is called three times to catch different flows/edge cases, I wonder if it would be good to abstract in a hook and I wonder if we should be using `domReady` somehow here?

File: packages/block-editor/src/components/iframe/index.js [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+		}
+	}
+
+	return (
+		<iframe
+			{ ...props }
+			style={ {
+				display: 'block',
+				width: '100%',
+				height: sizes.height + 'px',
```
Comment (youknowriad) [inline packages/block-editor/src/components/iframe/index.js:205]: Is a fixed height necessary for iframes?

File: packages/block-editor/src/components/use-resize-canvas/index.js [hunk 17]
```diff
@@ -11,11 +11,15 @@ import { default as useSimulatedMediaQuery } from '../../components/use-simulate
 /**
  * Function to resize the editor window.
  *
- * @param {string} deviceType Used for determining the size of the container (e.g. Desktop, Tablet, Mobile)
+ * @param {string}  deviceType Used for determining the size of the container (e.g. Desktop, Tablet, Mobile)
+ * @param {boolean} __unstableDisableSimulate
```
Comment (youknowriad) [inline packages/block-editor/src/components/use-resize-canvas/index.js:15]: could use docs maybe.

File: packages/components/src/popover/index.js [hunk 18]
```diff
@@ -363,42 +406,68 @@ const Popover = ( {
...
+				anchorDocument.defaultView.removeEventListener(
+					'resize',
+					refresh
+				);
+				anchorDocument.defaultView.removeEventListener(
+					'scroll',
+					refresh,
+					true
+				);
+			}
```
Comment (youknowriad) [inline packages/components/src/popover/index.js:470]: The changes in popover are a bit hard to understand, should we add some inline docs? What problems did  you face that required the changes here?

File: packages/e2e-test-utils/src/canvas.js [hunk 19]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Gets the editor canvas frame.
+ */
+export function canvas() {
```
Comment (youknowriad) [inline packages/e2e-test-utils/src/canvas.js:4]: maybe getEditorCanvas or something more descriptive

File: packages/e2e-test-utils/src/show-block-toolbar.js [hunk 20]
```diff
@@ -3,7 +3,9 @@
  * Call this function to reveal it.
  */
 export async function showBlockToolbar() {
+	const content = await page.$( '.interface-interface-skeleton__content' );
+	const { x, y } = await content.boundingBox();
 	// Move the mouse to disable the isTyping mode
-	await page.mouse.move( 50, 50 );
-	await page.mouse.move( 100, 100 );
+	await page.mouse.move( x + 50, y + 50 );
+	await page.mouse.move( x + 100, y + 100 );
```
Comment (youknowriad) [inline packages/e2e-test-utils/src/show-block-toolbar.js:10]: Why this change was necessary?

File: packages/e2e-tests/specs/experiments/multi-entity-editing.test.js [hunk 21]
```diff
@@ -50,8 +51,8 @@ const createTemplatePart = async (
 	await page.keyboard.type( templatePartName );
 };
 
-const editTemplatePart = async ( textToAdd, isNested = false ) => {
-	await page.click(
+const editTemplatePart = async ( textToAdd, isNested = false, p = page ) => {
```
Comment (youknowriad) [inline packages/e2e-tests/specs/experiments/multi-entity-editing.test.js:54]: this extra arg is useless right? I'm not saying where it's being passed.

File: packages/edit-site/src/components/block-editor/index.js [hunk 22]
```diff
@@ -41,9 +63,8 @@ export default function BlockEditor( { setIsInserterOpen } ) {
 		templateType
 	);
 	const { setPage } = useDispatch( 'core/edit-site' );
-	const ref = useRef();
 
-	useBlockSelectionClearer( ref );
+	const resizedCanvasStyles = useResizeCanvas( deviceType, true );
```
Comment (youknowriad) [inline packages/edit-site/src/components/block-editor/index.js:67]: Why is this extra hook needed, it seems iframe specific right? Can't we embed it in the iframe component?

**Review by youknowriad — COMMENTED** [review]
Awesome work here, the fact that the code doesn't look that complex where it's obviously a gigantic task says a lot.

File: lib/client-assets.php (hunk 8 above)
Comment (ellatrix) [reply lib/client-assets.php:699]: Yeah, I can look into that

File: packages/block-editor/src/components/block-list/block-selection-button.js (hunk 10 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block-selection-button.js:231]: It's navigation mode behaviour. The fallback is for when you're in an iframe and there is no next element to tab to. I should probably find a more elegant solution for this.

File: packages/block-editor/src/components/iframe/index.js (hunk 11 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:51]: That's correct! We could do that, yes.

File: packages/block-editor/src/components/iframe/index.js (hunk 12 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:63]: It's only for the drop zone and keyboard shortcuts. I'm confident we can remove the need for both, but keyboard shortcuts might be a bit of a larger refactor.

File: packages/block-editor/src/components/iframe/index.js (hunk 13 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:140]: I think I'm going to remove this iframe resizing. The reason I did that was for meta boxes in the post editor. We'll just need to find a better way to do that and let the iframe be scrollable instead of expanding it. @jasmussen noticed a bug with it above.

File: packages/block-editor/src/components/iframe/index.js (hunk 16 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:205]: It's being auto resized, but I think I'll remove this behaviour.

File: packages/components/src/popover/index.js (hunk 18 above)
Comment (ellatrix) [reply packages/components/src/popover/index.js:470]: Popover positions need to change when either the iframe or the parent window scrolls or resizes, because the anchor is within an iframe and the popover is outside of it.

File: packages/e2e-test-utils/src/canvas.js (hunk 19 above)
Comment (ellatrix) [reply packages/e2e-test-utils/src/canvas.js:4]: I just wanted to keep it small like `page` so it doesn't trigger reformatting in the many places we need to update e2e tests for the post editor later. See #21102.

File: packages/e2e-test-utils/src/show-block-toolbar.js (hunk 20 above)
Comment (ellatrix) [reply packages/e2e-test-utils/src/show-block-toolbar.js:10]: The rectangle of elements in the iframe will be relative to the iframe position. We need to offset that.

File: packages/edit-site/src/components/block-editor/index.js (hunk 22 above)
Comment (ellatrix) [reply packages/edit-site/src/components/block-editor/index.js:67]: It was already here. It's to resize the canvas for preview. Imo it's an implementation detail.

File: packages/block-editor/src/components/iframe/index.js (hunk 13 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js]: Removed this now.

File: packages/block-editor/src/components/iframe/index.js (hunk 15 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js]: I simplified this now. I don't think that `domReady` is good here since the goal is different. For the iframe, we're looking for a usable document as quickly as possible. Only Firefox is a bit slow here, where it's only ready on the `load` event. In other browsers the document is ready as soon as the iframe mounts.

**Comment by ItsJonQ** [comment]
@ellatrix This is absolutely amazing 😍 ! I can't wait for this to land.

I took it for a spin locally. Some of the issues I noticed were reported above - apologies for duplicates.

### 📏 Heights are fighting

Screen Shot 2020-12-10 at 10 29 42 AM

There's a "battle of heights" between the `iFrame` (element), `iFrame` content, and editor interface that contains the iFrame (and potentially `window.document.body` as well).

Even with clever dynamic height calculations added to the iFrame (for example, `calc(100vh + ${heightOfTopBar})`), things aren't coordinating.

Layout styles, especially height related stuff, is always really tricky. This setup is basically attempting to do this is "Very Hard mode".

I've noticed "jumpiness" when interacting with the editor. I think it comes back to the mismatched heights.

I'm not sure how to resolve this yet.

It will take some time poking and playing with various height styles (and perhaps dynamic calculation) to make all of the layouts happy.

### 🐑 Stylesheet Cloning

The technique you have is 👍 . It's something I've done in the past. It's surprising how simple (and effective) that technique is (the CSS engine is pretty great).

I don't think this would be a problem (at all). But... if... IF... there are dynamic styles (either stylesheets or inline styles) that are injected AFTER the fact (during the lifecycle of the editor's session), those won't be cloned over.

The resolve this, we can create a MutationObserver to transfer any newly injected stylesheets

(Again, this is only if there is a problem).

### 🎨 Theme vs Editor

It looks like the iFrame only encompasses the Canvas. All the things in the sidebar, topbar, etc... are outside of the iFrame, which is okay. Unfortunately, this subjects them to whatever stylesheets that may load.

For example, if a theme has global styles, like `:root { font-weight: 300; }`, the editor's UI will be affected.

Screen Shot 2020-12-10 at 10 30 32 AM

Fortunately, this won't be a problem when we successfully roll-out the [upgraded Component System](https://github.com/WordPress/gutenberg/issues/27484), as the style system is designed to be as stable as possible, regardless of environment (with zero side effects).

---

That's all I've got! Hopefully this is helpful.

**Comment by ellatrix** [comment]
@ItsJonQ Just wondering, did you test the branch before or after the removal of auto resizing?

**Comment by ellatrix** [comment]
> The technique you have is 👍 . It's something I've done in the past. It's surprising how simple (and effective) that technique is (the CSS engine is pretty great).

This is only meant to be a compatibility layer of stylesheets loaded on the server side. Ideally editor style sheets need to be set the official way, not just enqueuing the styles on the page.

> It looks like the iFrame only encompasses the Canvas. All the things in the sidebar, topbar, etc... are outside of the iFrame, which is okay. Unfortunately, this subjects them to whatever stylesheets that may load.

I'm not sure I'm following. The iframe is meant to only contain content, not the rest of the editor. The styles for the content and editor UI are different.

File: packages/e2e-test-utils/src/show-block-toolbar.js (hunk 20 above)
Comment (youknowriad) [reply packages/e2e-test-utils/src/show-block-toolbar.js:10]: But the idea here is just to move the mouse randomly to trigger the block toolbar, not sure we need precise position?

**Comment by ItsJonQ** [comment]
> Just wondering, did you test the branch before or after the removal of auto resizing?

I'm not sure 🤔 . The latest commit I have for my local branch is [Simplify](https://github.com/WordPress/gutenberg/pull/25775/commits/7279dce588b7c9195b9dce64cffb62740d78ecc5)

@ellatrix 

> I'm not sure I'm following. The iframe is meant to only contain content, not the rest of the editor. The styles for the content and editor UI are different.

Ah! My apologies. My comment was unrelated to the iFrame content work. It was more of an observation 😊

**Comment by jasmussen** [comment]
Just tested the latest version which removes the auto resizing:

- There's still a small resize handle on the appender
- There's still a horizontal scrollbar in some cases

But both of those are existing issues that are only uncovered by the iframe. The resize handle removal was inherited from somewhere and that CSS should be moved to the appender instead. And the horizontal scrollbar is created by this:

```
.block-editor-block-list__layout.has-overlay::after {
    content: "";
    position: absolute;
    top: -14px;
    right: -14px;
    bottom: -14px;
    left: -14px;
    z-index: 60;
}
```

those -14px are responsible.

But the removal of the autosizing makes this go from feeling sensitive, to feeling totally solid. I'm over the moon about this, and I think we can land it 🎉  — it's my strong impression that all the remaining issues are best solved as small separate PRs, and that the entire block editor will be better for it. Amazing work.

File: lib/client-assets.php (hunk 8 above)
Comment (ellatrix) [reply lib/client-assets.php:699]: I tried to do this, but the styles are not available yet at the time when the editor scripts are registered.

File: packages/e2e-test-utils/src/show-block-toolbar.js (hunk 20 above)
Comment (ellatrix) [reply packages/e2e-test-utils/src/show-block-toolbar.js:10]: With the iframe, the mouse needs to be over the canvas to trigger the block toolbar. Does that not make sense? I can change it so it works for the main window too.

File: packages/block-editor/src/components/iframe/index.js (hunk 14 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js]: Removed the alignment default style. The other one is specific for the iframe context.

File: packages/block-editor/src/components/iframe/index.js (hunk 11 above)
Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:58]: There's still some of our own stylesheets left as well. Maybe we should attempt this in a follow-up?

File: packages/e2e-test-utils/src/show-block-toolbar.js (hunk 20 above)
Comment (ellatrix) [reply packages/e2e-test-utils/src/show-block-toolbar.js]: I restored the behaviour.

File: packages/e2e-tests/specs/experiments/multi-entity-editing.test.js (hunk 21 above)
Comment (ellatrix) [reply packages/e2e-tests/specs/experiments/multi-entity-editing.test.js]: Yes, removed it.

File: packages/components/src/popover/index.js (hunk 18 above)
Comment (ellatrix) [reply packages/components/src/popover/index.js:472]: Added a comment.

File: packages/block-editor/src/components/iframe/index.js (hunk 11 above)
Comment (youknowriad) [reply packages/block-editor/src/components/iframe/index.js:58]: yes, definitely, what are our own stylesheets that falls here?

File: packages/block-editor/src/components/observe-typing/index.js [hunk 23]
```diff
@@ -43,12 +43,62 @@ function isKeyDownEligibleForStartTyping( event ) {
 	return ! shiftKey && KEY_DOWN_ELIGIBLE_KEY_CODES.has( keyCode );
 }
 
+export function useMouseMoveTypingReset( ref ) {
```
Comment (youknowriad) [inline packages/block-editor/src/components/observe-typing/index.js:46]: Would you mind adding docs explaining what this does especially since it's exported :)

File: packages/block-editor/src/components/use-resize-canvas/index.js [hunk 24]
```diff
@@ -11,11 +11,15 @@ import { default as useSimulatedMediaQuery } from '../../components/use-simulate
...
- * @param {string} deviceType Used for determining the size of the container (e.g. Desktop, Tablet, Mobile)
+ * @param {string}  deviceType                  Used for determining the size of the container (e.g. Desktop, Tablet, Mobile)
+ * @param {boolean} __unstableDisableSimulation Whether to disable media query simulation.
  *
  * @return {Object} Inline styles to be added to resizable container.
  */
-export default function useResizeCanvas( deviceType ) {
+export default function useResizeCanvas(
+	deviceType,
+	__unstableDisableSimulation
```
Comment (youknowriad) [inline packages/block-editor/src/components/use-resize-canvas/index.js:21]: I  think at some point, when the iframe becomes more the "default" we should consider removing this entirely. (no responsive support for non-iframes)

Comment (ellatrix) [reply packages/block-editor/src/components/use-resize-canvas/index.js:21]: Agreed, but for now it's still used by the post editor.

File: lib/client-assets.php (hunk 8 above)
Comment (youknowriad) [reply lib/client-assets.php:700]: I  was not suggesting at the script registration, I was more suggesting `block_editor_settings` filter for post editor and `gutenberg_edit_site_init` for the site editor.

File: packages/block-editor/src/components/observe-typing/index.js (hunk 23 above)
Comment (ellatrix) [reply packages/block-editor/src/components/observe-typing/index.js:54]: Of course, sorry about that. Added docs for `useTyingObserver` too.

File: lib/client-assets.php (hunk 8 above)
Comment (ellatrix) [reply lib/client-assets.php:700]: That's what I mean. The `block_editor_settings` filter is run at script registration, which is too early. The settings are passed as an inline script initialising the editor.

File: packages/block-editor/src/components/iframe/index.js [hunk 25]
```diff
@@ -148,27 +148,35 @@ function Iframe( { contentRef, children, head, ...props }, ref ) {
...
+		node.addEventListener( 'load', () => {
+			setDocumentIfReady();
+		} );
 	}, [] );
 
 	return (
 		<iframe
 			{ ...props }
-			ref={ mergeRefs( [ ref, setRef ] ) }
+			ref={ useCallback( mergeRefs( [ ref, setRef ] ), [] ) }
```
Comment (youknowriad) [inline packages/block-editor/src/components/iframe/index.js:179]: I'm pretty sure, the useCallback here is useless.

Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:179]: It's not, see #27686.

Comment (youknowriad) [reply packages/block-editor/src/components/iframe/index.js:179]: Interesting :)

**Comment by jasmussen** [comment]
💪

File: packages/e2e-test-utils/src/canvas.js (hunk 19 above)
Comment (talldan) [inline packages/e2e-test-utils/src/canvas.js:4]: Not sure the reason behind naming it this way was worthwhile.

Seeing it used in a test as `canvas()`, it looks really bizarre.

Comment (ellatrix) [reply packages/e2e-test-utils/src/canvas.js:4]: Any suggestions? This will be used a lot in all tests when the iframe will be used for the post editor, so it would be good if it's short like `page` so it doesn't trigger code reformatting for thousands of lines.
